csharp/src/Ice/SlicSocket.cs

```diff
@@ -543,6 +512,62 @@ internal void TraceTransportFrame(string prefix, SlicDefinitions.FrameType type,
             Endpoint.Communicator.Logger.Trace(TraceLevels.TransportCategory, s.ToString());
         }
 
+        private protected override SocketStream CreateControlStream() =>
+            // We make sure to allocate the ID for the control stream right away. Otherwise, it would be considered
+            // like a regular stream and if serialization is enabled, it would acquire the semaphore.
+            new SlicStream(AllocateId(false), this);
+
+        private void ReadParameters(ReadOnlyMemory<byte> data, InputStream istr)
+        {
+            int dictionarySize = istr.ReadSize();
+            var parameters = new Dictionary<int, ReadOnlyMemory<byte>>(dictionarySize);
+            for (int i = 0; i < dictionarySize; ++i)
+            {
+                int key = istr.ReadVarInt();
+                int entrySize = istr.ReadSize();
+                if (!parameters.TryAdd(key, data.Slice(istr.Pos, entrySize)))
```

Comment:
I find this code inelegant. istr is a stream over data: we should be able read ROM from the InputStream's underlying buffer.

Reply:
I just copied the code for reading the binary context. I assumed we didn't want to add an API on InputStream that returned a ROM struct that referred to the internal stream buffer and that it was best to let the caller deal with the creation of the ROM.

So what do you suggest? Add `ReadBinaryContextEntry` that return a `(int, ROM)` pair without copying the data?

Reply:
Yes, sounds good.